File: 1798-max-number-of-k-sum-pairs/1798-max-number-of-k-sum-pairs.py
class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        count = 0
        keep = defaultdict(int)

        for n in nums:
            other = k - n
            if keep[other] > 0:
                keep[other] -= 1
                count += 1
            else:
                keep[n] += 1
        
        return count


        # count = 0
        # nums.sort()
        # i, j = 0, len(nums) - 1

        # while i < j:
        #     s = nums[i] + nums[j]
        #     if s > k:
        #         j -= 1
        #     elif s < k:
        #         i += 1
        #     else:
        #         i += 1
        #         j -= 1
        #         count += 1

        # return count 

File: 1798-max-number-of-k-sum-pairs/1798-max-number-of-k-sum-pairs.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maxOperations = function(nums, k) {

    const map = new Map();
    let res = 0;

    for (const n of nums) {
        const e = k - n;

        const val = map.get(e);

        if (val !== undefined && val > 0) {
            res++;
            map.set(e, val - 1);
        } else {
            map.set(n, (map.get(n) || 0) + 1);
        }
    }

    return res;
    
};